src/assignments/benson/advent_2018/task2.clj

```diff
@@ -0,0 +1,71 @@
+(ns assignments.benson.advent-2018.task2)
+
+;;;; ------------------ Part 1 ------------------
+
+(def task2input (slurp "src/assignments/benson/advent_2018/task2input"))
+
+(def ids (clojure.string/split-lines task2input))
+
+(defn match-2 
+  "returns if id has 2 matching letters"
+  [id]
+  (contains? (set (vals (frequencies id))) 2))
+
+(defn match-3 
+  "returns if id has 3 matching letters"
+  [id]
+  (contains? (set (vals (frequencies id))) 3))
+
+(assert (match-2 "abcdec"))
+(assert (not (match-2 "abcdef")))
+(assert (match-3 "accdec"))
+(assert (not (match-3 "abcdef")))
+(assert (and (match-2 "aacbbb") (match-3 "aacbbb")))
+
+(defn calculate-checksum 
+  "Returns the checksum calculated by multiplying the count of ID's with 2 matching letters
+   and ID's with 3 matching letters"
+  [ids]
+  (loop [[curr-id & rest-ids] ids
+         two-matches 0
+         three-matches 0]
+    (if curr-id
+      (recur rest-ids
+             (if (match-2 curr-id) (inc two-matches) two-matches) 
+             (if (match-3 curr-id) (inc three-matches) three-matches))
+      (* two-matches three-matches))))
+
+; (calculate-checksum ids) -> 4980
+
+;;;; ------------------ Part 2 ------------------
+
+(def id-length (count (first ids)))
+
+(defn split-id [id] (re-seq #"\w" id))
+
+(defn non-matches
+  "Returns how many letters in ID-1 don't match with its corresponding 
+   letter in the other ID-2"
+  [id-1 id-2]
+  (let [split-id-1 (split-id id-1)
+        split-id-2 (split-id id-2)]
+    (count (filter false? (map = split-id-1 split-id-2)))))
+
+(assert (= 3 (non-matches "abcdefg" "abcOOOg")))
+
+(defn get-similar-ids
+  "Returns the ID's with only 1 character difference"
+  [ids]
+  (let [similar-ids (promise)]
+    (doseq [id-1 ids
+            id-2 ids]
+      (future (if (= 1 (non-matches id-1 id-2))
```

Comment:
Ouch.  This presumably tries to do some simple computation?  Don't use futures for that!!! :-)

I don't know what the function is trying to compute; can you state the problem definition, or show me an assert of expected input/output?  [I've not seen this problem before, I think].  

These are really good problems for you to work on!  Keep going!

Reply:
This problem needed me to compare 1000 ids with each other (1000 * 999 comparisons) which makes the program block / freeze for several seconds. I figured I try out futures and deliver to try and make the problem not block and concurrently calculate just for some practice.

Reply:
Ah, OK. 

In that case, the clearer way to do that is to have a function which hammers the cpu, and wrap that in a future,
but not intermingle the two ideas.  Make sense?

How about something like this?
```
(defn near-miss [a b] (= 1 (non-matches a b)))

(defn all-distinct-pairs [coll]
  (for [x coll
        y coll
        :when (not= x y)]
    [x y]))

(defn get-similar-ids
  "Returns the ID's with only 1 character difference"
  [ids]
  (filter near-miss (all-distinct-pairs ids)))

(defn test-it [ids]
  (let [result (atom nil)]
    (future (reset! result (get-similar-ids ids)))
    result))

;; And let caller keep dereferencing result at leisure until
;; results are in there, 100% cpu usage, repl not blocked.
```

Reply:
I didn't think about chucking everything into 1 future.
Thanks for the idea!